docs_src/design/adr/system-management/0006-Metrics-ControlPlaneEvents.md

```diff
@@ -0,0 +1,226 @@
+
+The collection and dissemination of metric data and CPE will require service level instrumentation relevant to capture and send data about relevant EdgeX operations.  EdgeX does not currently offer any service instrumentation.
+
+### Metric and CPE Consumption/Use
+
+At this time, EdgeX will make metric data and CPE available to other subscribing 3rd party applications and systems, but will not necessarily consume or use this information itself.  
+
+In the future, EdgeX may consume its own metric data or CPE.  For example, EdgeX may, in the future,
+
+- use a metric on the number of EdgeX events being sent to core data as the means to throttle back device data collection
+- use a configuration change CPE to automatically restart a service.
+
+In the future, EdgeX application services may optionally subscribe to metrics and CPE messages (by attaching to the appropriate message pipe).  Thus allowing additional filtering, transformation, endpoint control of metric and CPE data.
+
+- At the point where this feature is supported, consideration would need to be made as to whether all events (sensor reading messages and metric/CPE messages go through the same application services)
+
+At this time, EdgeX will not persist the metric or CPE data (except as it may be retained as part of message delivery).  Consumers of metric or CPE data are responsible for persisting the data if needed.  Persistence of metric or CPE information may be considered in the future.
+
+In general, EdgeX metrics and CPE are meant to provide external applications and systems better information about what is happening "inside" EdgeX and the devices with which it communicates.
+
+### Requirements
+
+- Services will push specified metrics collected for that service to a specified (by configuration) message endpoint (MQTT topic, 0MQ topic or other EdgeX supported messaging implementation)
+  - Use and configuration to target a secure endpoint (MQTTS) must be provided.  As exemplified by application services, there must be the option to use "insecure secrets" or Vault-backed security services when a secure endpoint is desired.
+- Services will push all CPE detected for that service to a specified (by configuration) message endpoint.
+  - The endpoint for metric and CPE may be the same or different at the user's discretion.
+  - Use and configuration to target a secure endpoint (MQTTS) must be provided.  As exemplified by application services, there must be the option to use "insecure secrets" or Vault-backed security services when a secure endpoint is desired.
+- All services must document what metrics and CPE they offer.
+- Services must provide REST endpoints that respond with what metrics and CPE the service offers (and whether that metric or CPE is currently `on` or `off`).
+  - Proposed endpoint format:  /api/v2/metrics and /api/v2/cpe
+- Services will have configuration which allows EdgeX system managers to select which metrics and CPE are turned `on` or `off`.

```

Comment:
if the state of contain `on` and `off`, I suggest using bool (`true` or `false`)
on the other hand, if we use string, we can define more state, such as `off`, `basic`, and `trace` (maybe some better naming).
metrics collection might consume more resource and it would slow down the performance a lot in a low power machine.  we can set different level of metrics or CPE collection for different situation.

Reply:
I think we want the ability to turn Metrics and/or CPE on or off and when on which metrics/CPE we want to collect/send. For metrics we may need to categorize them for enabling/disabling since there will be many, many instances. CPE will likely just be a few, so do we care to filter them when CPE is on?

Reply:
Agree - control must be more than just on or off, but a level like in logging.  I've added some additional wording to this area and proposed a `level`.